File: CreationalDesignPatterns/Java/AbstracFActory.java
// Abstract Product
interface Button {
    void render();
}

interface Checkbox {
    void render();
}

// Concrete Product
class WindowsButton implements Button {
    public void render() {
        System.out.println("Rendering Windows Button");
    }
}

class MacButton implements Button {
    public void render() {
        System.out.println("Rendering Mac Button");
    }
}

class WindowsCheckbox implements Checkbox {
    public void render() {
        System.out.println("Rendering Windows Checkbox");
    }
}

class MacCheckbox implements Checkbox {
    public void render() {
        System.out.println("Rendering Mac Checkbox");
    }
}

// Abstract Factory
interface GUIFactory {
    Button createButton();
    Checkbox createCheckbox();
}

// Concrete Factory
class WindowsFactory implements GUIFactory {
    public Button createButton() {
        return new WindowsButton();
    }
    public Checkbox createCheckbox() {
        return new WindowsCheckbox();
    }
}

class MacFactory implements GUIFactory {
    public Button createButton() {
        return new MacButton();
    }
    public Checkbox createCheckbox() {
        return new MacCheckbox();
    }
}

// Client Code
public class Main {
    public static void main(String[] args) {
        GUIFactory factory;

        // Windows UI তৈরি
        factory = new WindowsFactory();
        Button windowsButton = factory.createButton();
        Checkbox windowsCheckbox = factory.createCheckbox();
        windowsButton.render(); // Output: Rendering Windows Button
        windowsCheckbox.render(); // Output: Rendering Windows Checkbox

        // Mac UI তৈরি
        factory = new MacFactory();
        Button macButton = factory.createButton();
        Checkbox macCheckbox = factory.createCheckbox();
        macButton.render(); // Output: Rendering Mac Button
        macCheckbox.render(); // Output: Rendering Mac Checkbox
    }
}

File: CreationalDesignPatterns/JavaScript/AbstracFActory.js
// Abstract Product
class Button {
    render() {
        throw new Error("This method should be overridden!");
    }
}

class Checkbox {
    render() {
        throw new Error("This method should be overridden!");
    }
}

// Concrete Product
class WindowsButton extends Button {
    render() {
        console.log("Rendering Windows Button");
    }
}

class MacButton extends Button {
    render() {
        console.log("Rendering Mac Button");
    }
}

class WindowsCheckbox extends Checkbox {
    render() {
        console.log("Rendering Windows Checkbox");
    }
}

class MacCheckbox extends Checkbox {
    render() {
        console.log("Rendering Mac Checkbox");
    }
}

// Abstract Factory
class GUIFactory {
    createButton() {
        throw new Error("This method should be overridden!");
    }
    createCheckbox() {
        throw new Error("This method should be overridden!");
    }
}

// Concrete Factory
class WindowsFactory extends GUIFactory {
    createButton() {
        return new WindowsButton();
    }
    createCheckbox() {
        return new WindowsCheckbox();
    }
}

class MacFactory extends GUIFactory {
    createButton() {
        return new MacButton();
    }
    createCheckbox() {
        return new MacCheckbox();
    }
}

// Client Code
function createUI(factory) {
    const button = factory.createButton();
    const checkbox = factory.createCheckbox();

    button.render(); // ফ্যাক্টরি অনুযায়ী বাটন রেন্ডার হবে
    checkbox.render(); // ফ্যাক্টরি অনুযায়ী চেকবক্স রেন্ডার হবে
}

// Windows UI তৈরি
const windowsFactory = new WindowsFactory();
createUI(windowsFactory);

// Mac UI তৈরি
const macFactory = new MacFactory();
createUI(macFactory);
